### PR TITLE
Add New ExecOutput Function To Support Custom Writers

### PR DESCRIPTION
**Pull request by ihcsim**
## Change Overview

This PR introduces a new `kube.ExecOutput()` library function that can write `kube.Exec()` outputs to custom writers, allowing callers to provide either buffered writers or async (pipe) writers. This new function delegates the writes operation to a new `format.Writer` instance that formats the outputs to match the Kanister-style logs, before writing them to the custom writers.

The `KubeExec` Kanister function has been updated to use `kube.ExecOutput()`, with `os.Stdout` and `os.Stderr`. Other callers in the `pkg/app` and `pkg/function` packages remain untouched. By offering both `kube.Exec()` and `kube.ExecAsync()`, downstream has the flexibility to decide which output mechanism to use, based on their use cases.

## Pull request type

Please check the type of change your PR introduces:
- [ ] :construction: Work in Progress
- [ ] :rainbow: Refactoring (no functional changes, no api changes)
- [ ] :hamster: Trivial/Minor
- [ ] :bug: Bugfix
- [x] :sunflower: Feature
- [ ] :world_map: Documentation
- [ ] :robot: Test

## Issues

Follow-up of:
- https://github.com/kanisterio/kanister/pull/1194
- https://github.com/kanisterio/kanister/pull/1037

## Test Plan

Deploy this `timer` pod that has a script to tail the current datetime at one second interval:

```sh
cat <<EOF | k apply -f -
apiVersion: v1
kind: Pod
metadata:
  name: timer
spec:
  containers:
  - name: timer
    image: ubuntu
    command:
    - /bin/sh
    - "-c"
    - |
      cat <<EOF > /bin/tail_datetime.sh
      for i in \\\$(seq 0 20)
      do
        date +"%m-%d-%Y %H:%M:%S"
        sleep 1s
      done
      EOF

      chmod +x /bin/tail_datetime.sh
      sleep 7200s
EOF
```

Deploy this blueprint that uses `KubeExec` to invoke the remote script in the `timer` pod:

```sh
cat <<EOF | k apply -f -
apiVersion: cr.kanister.io/v1alpha1
kind: Blueprint
metadata:
  name: timer
  namespace: kanister
actions:
  tail:
    phases:
    - func: KubeExec
      name: tail
      args:
        namespace: default
        pod: timer
        container: timer
        command:
        - sh
        - "-c"
        - /bin/tail_datetime.sh
EOF
```

Creates the actionset to trigger the function call:
```sh
cat <<EOF | k create -f -
apiVersion: cr.kanister.io/v1alpha1
kind: ActionSet
metadata:
  generateName: timer-
  namespace: kanister
spec:
  actions:
  - name: tail
    blueprint: timer
    object:
      kind: Namespace
      name: default
EOF
```

Observe the datetime being streamed to the Kanister operator's logs:

```sh
$ k -n kanister logs -f kanister-kanister-operator-5964486449-xqv66
#...

{"Container":"timer","File":"pkg/log/log.go","Function":"github.com/kanisterio/kanister/pkg/log.PrintTo","Line":174,"Out":"03-25-2022 05:52:49","Pod":"timer","cluster_name":"8224a31e-876b-4ef2-98c2-ae9cbc5e96b0","hostname":"kanister-kanister-operator-5964486449-xqv66","level":"info","msg":"action update","time":"2022-03-25T05:52:49.060833604Z"}
{"Container":"timer","File":"pkg/log/log.go","Function":"github.com/kanisterio/kanister/pkg/log.PrintTo","Line":174,"Out":"03-25-2022 05:52:50","Pod":"timer","cluster_name":"8224a31e-876b-4ef2-98c2-ae9cbc5e96b0","hostname":"kanister-kanister-operator-5964486449-xqv66","level":"info","msg":"action update","time":"2022-03-25T05:52:50.063202873Z"}
{"ActionSet":"timer-2pkxw","File":"pkg/controller/controller.go","Function":"github.com/kanisterio/kanister/pkg/controller.(*Controller).logAndSuccessEvent","Line":510,"Phase":"tail","cluster_name":"8224a31e-876b-4ef2-98c2-ae9cbc5e96b0","hostname":"kanister-kanister-operator-5964486449-xqv66","level":"info","msg":"Completed phase tail","time":"2022-03-25T05:52:51.077562462Z"}
{"File":"pkg/controller/controller.go","Function":"github.com/kanisterio/kanister/pkg/controller.(*Controller).logAndSuccessEvent","Line":510,"cluster_name":"8224a31e-876b-4ef2-98c2-ae9cbc5e96b0","hostname":"kanister-kanister-operator-5964486449-xqv66","level":"info","msg":"Updated ActionSet 'timer-2pkxw' Status-\u003ecomplete","time":"2022-03-25T05:52:51.081930515Z"}
```

- [ ] :muscle: Manual
- [ ] :zap: Unit test
- [ ] :green_heart: E2E

### CODE REVIEW

File: pkg/function/kube_exec.go [hunk 1]
```diff
@@ -91,15 +91,11 @@ func (kef *kubeExecFunc) Exec(ctx context.Context, tp param.TemplateParams, args
...
 	}
-	stdout, stderr, err := kube.Exec(cli, namespace, pod, container, cmd, nil)
-	format.LogWithCtx(ctx, pod, container, stdout)
-	format.LogWithCtx(ctx, pod, container, stderr)
-	if err != nil {
+	if err := kube.ExecOutput(cli, namespace, pod, container, cmd, nil, os.Stdout, os.Stderr); err != nil {
 		return nil, err
 	}
 
-	out, err := parseLogAndCreateOutput(stdout)
```
Comment (pavannd1) [inline pkg/function/kube_exec.go:101]: I think this is still required if used by the ActionSet to set output artifacts

File: pkg/kube/exec.go [hunk 2]
```diff
@@ -75,19 +121,22 @@ func ExecWithOptions(kubeCli kubernetes.Interface, options ExecOptions) (string,
...
-	if err != nil {
-		return "", "", err
-	}
+	errCh := make(chan error, 1)
+	go func() {
+		err := execute("POST", req.URL(), config,
+			options.Stdin,
+			options.Stdout,
+			options.Stderr,
+			tty)
```
Comment (pavannd1) [inline pkg/kube/exec.go:135]:
```suggestion
		err := execute(
			"POST",
			req.URL(),
			config,
			options.Stdin,
			options.Stdout,
			options.Stderr,
			tty,
		)
```
[nit] I like the formatting here - https://github.com/thanos-io/thanos/blob/main/docs/contributing/coding-style-guide.md#only-two-ways-of-formatting-functionsmethods
Improves readability 😄

File: pkg/function/kube_exec.go (hunk 1 above)
Comment (ihcsim) [reply pkg/function/kube_exec.go:101]: Good catch; will fix.

Comment (ihcsim) [reply pkg/function/kube_exec.go:101]: Fixed in 17c673ff35f024b5757c22b2894002278bde4839. I used this blueprint to emulate a phase output:

```sh
apiVersion: cr.kanister.io/v1alpha1
kind: Blueprint
metadata:
  name: timer
  namespace: kanister
actions:
  tail:
    outputArtifacts:
      output:
        keyValue:
          foo: "{{ .Phases.tail.Output.foo }}"
    phases:
    - func: KubeExec
      name: tail
      args:
        namespace: default
        pod: timer
        container: timer
        command:
        - sh
        - "-c"
        - |
         /bin/tail_datetime.sh
         echo '###Phase-output###: {"key":"foo","value":"bar"}'
```

**Review by pavannd1 — APPROVED** [review]
LGTM

**Pull request merged**